src/Elastic.Apm/Report/PayloadSenderV2.cs

```diff
@@ -0,0 +1,250 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using Elastic.Apm.Api;
+using Elastic.Apm.Config;
+using Elastic.Apm.Logging;
+using Elastic.Apm.Model.Payload;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Serialization;
+
+namespace Elastic.Apm.Report
+{
+	/// <summary>
+	/// Responsible for sending the data to the server. Implements Intake V2.
+	/// Each instance creates its own thread to do the work. Therefore, instances should be reused if possible.
+	/// </summary>
+	internal class PayloadSenderV2 : IPayloadSender
+	{
+		private static readonly int DnsTimeout = (int)TimeSpan.FromMinutes(1).TotalMilliseconds;
+
+		private readonly Task _creation;
+
+		private readonly BatchBlock<object> _eventQueue =
+			new BatchBlock<object>(20, new GroupingDataflowBlockOptions
+				{ BoundedCapacity = 1_000_000 });
+
+		private readonly HttpClient _httpClient;
+		private readonly ScopedLogger _logger;
+
+		private readonly Service _service;
+
+		private CancellationTokenSource _batchBlockReceiveAsyncCts;
+
+		private readonly JsonSerializerSettings _settings;
+
+		private readonly SingleThreadTaskScheduler _singleThreadTaskScheduler = new SingleThreadTaskScheduler(CancellationToken.None);
+
+		public PayloadSenderV2(IApmLogger logger, IConfigurationReader configurationReader, Service service, HttpMessageHandler handler = null)
+		{
+			_settings = new JsonSerializerSettings { ContractResolver = new CamelCasePropertyNamesContractResolver(), Formatting = Formatting.None };
+			_service = service;
+			_logger = logger?.Scoped(nameof(PayloadSenderV2));
+
```

Comment:
There is a feature that needs multiple Urls (don't know the details). Currently this feature is not implemented, but to avoid breaking changes later we already allow setting multiple Urls, we just use the first one currently.

Reply:
Yes, done: https://github.com/elastic/apm-agent-dotnet/issues/144